### PR TITLE
Milestone 2 - Ink Explorer Frontend

### PR DESCRIPTION
**Pull request by 0xslipk**
# Milestone Delivery Checklist

- [x] The [milestone-delivery-template.md](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/deliveries/milestone-delivery-template.md) has been copied and updated.
- [x] The [invoice form :pencil:](https://forms.gle/LSRr7PCjBpEbKGh89) has been filled out for this milestone.
- [x] This pull request is being made by the same account as the accepted application.
- [x] In case of acceptance, the payment will be transferred to the BTC/ETH payment address provided in the application.
- [x] The delivery is according to the [Guidelines for Milestone Deliverables](https://github.com/w3f/Grants-Program/blob/master/docs/milestone-deliverables-guidelines.md).

Link to the application pull request: https://github.com/w3f/Grants-Program/pull/1052.

### CODE REVIEW

**Comment by 0xCaso** [comment]
Hi @0xslipk! Thanks for the submission. I'm starting the evaluation, so I'll tag you if I need your help or something.

**Comment by 0xslipk** [comment]
> Hi @0xslipk! Thanks for the submission. I'm starting the evaluation, so I'll tag you if I need your help or something.

Awesome, let me know if you need help.

**Comment by 0xslipk** [comment]
@0xCaso we released a new [version](https://github.com/blockcoders/ink-substrate-explorer-frontend/releases/tag/v1.0.2), with a few fixes and support for Spanish speakers.

**Comment by 0xCaso** [comment]
Hi @0xslipk, I'm writing the evaluation right now. In the meantime, I want to let you know that the frontend and backend containers do not start, giving the following errors:
- Backend:
```
2022-11-22 11:33:19 > ink-substrate-explorer-api@1.0.4 start:prod
2022-11-22 11:33:19 > NODE_ENV=production node dist/src/main
2022-11-22 11:33:19
2022-11-22 11:33:24  - property DATABASE_SSL_CA has failed the following constraints: isString
2022-11-22 11:33:24
2022-11-22 11:33:24     at Object.validate (/usr/src/app/dist/src/env/env.validation.js:119:15)
2022-11-22 11:33:24     at Function.forRoot (/usr/src/app/node_modules/.pnpm/@nestjs+config@2.2.0_bkbhg4knmtesqnnagm7a2p5cuy/node_modules/@nestjs/config/dist/config.module.js:67:45)
2022-11-22 11:33:24     at Object.<anonymous> (/usr/src/app/dist/src/env/env.module.js:22:41)
2022-11-22 11:33:24     at Module._compile (node:internal/modules/cjs/loader:1105:14)
2022-11-22 11:33:24     at Object.Module._extensions..js (node:internal/modules/cjs/loader:1159:10)
2022-11-22 11:33:24     at Module.load (node:internal/modules/cjs/loader:981:32)
2022-11-22 11:33:24     at Function.Module._load (node:internal/modules/cjs/loader:822:12)
2022-11-22 11:33:24     at Module.require (node:internal/modules/cjs/loader:1005:19)
2022-11-22 11:33:24     at require (node:internal/modules/cjs/helpers:102:18)
2022-11-22 11:33:24     at Object.<anonymous> (/usr/src/app/dist/src/blocks/blocks.module.js:11:22)
```
- Frontend:
```
2022-11-22 11:34:40 Error: Could not find a production build in the '/usr/src/app/.next' directory. Try building your app with 'next build' before starting the production server. https://nextjs.org/docs/messages/production-start-no-build-id
2022-11-22 11:34:40     at NextNodeServer.getBuildId (/usr/src/app/node_modules/.pnpm/next@12.3.3_m5sxuueb27gk6ddc5gums6vtgq/node_modules/next/dist/server/next-server.js:169:23)
2022-11-22 11:34:40     at new Server (/usr/src/app/node_modules/.pnpm/next@12.3.3_m5sxuueb27gk6ddc5gums6vtgq/node_modules/next/dist/server/base-server.js:58:29)
2022-11-22 11:34:40     at new NextNodeServer (/usr/src/app/node_modules/.pnpm/next@12.3.3_m5sxuueb27gk6ddc5gums6vtgq/node_modules/next/dist/server/next-server.js:70:9)
2022-11-22 11:34:40     at NextServer.createServer (/usr/src/app/node_modules/.pnpm/next@12.3.3_m5sxuueb27gk6ddc5gums6vtgq/node_modules/next/dist/server/next.js:140:16)
2022-11-22 11:34:40     at async /usr/src/app/node_modules/.pnpm/next@12.3.3_m5sxuueb27gk6ddc5gums6vtgq/node_modules/next/dist/server/next.js:149:31
```

**Review by 0xCaso — CHANGES_REQUESTED** [review]
Hi @0xslipk, I'm writing the evaluation right now. In the meantime, I want to let you know that the frontend and backend containers do not start, giving the following errors:
- Backend:
```
2022-11-22 11:33:19 > ink-substrate-explorer-api@1.0.4 start:prod
2022-11-22 11:33:19 > NODE_ENV=production node dist/src/main
2022-11-22 11:33:19
2022-11-22 11:33:24  - property DATABASE_SSL_CA has failed the following constraints: isString
2022-11-22 11:33:24
2022-11-22 11:33:24     at Object.validate (/usr/src/app/dist/src/env/env.validation.js:119:15)
2022-11-22 11:33:24     at Function.forRoot (/usr/src/app/node_modules/.pnpm/@nestjs+config@2.2.0_bkbhg4knmtesqnnagm7a2p5cuy/node_modules/@nestjs/config/dist/config.module.js:67:45)
2022-11-22 11:33:24     at Object.<anonymous> (/usr/src/app/dist/src/env/env.module.js:22:41)
2022-11-22 11:33:24     at Module._compile (node:internal/modules/cjs/loader:1105:14)
2022-11-22 11:33:24     at Object.Module._extensions..js (node:internal/modules/cjs/loader:1159:10)
2022-11-22 11:33:24     at Module.load (node:internal/modules/cjs/loader:981:32)
2022-11-22 11:33:24     at Function.Module._load (node:internal/modules/cjs/loader:822:12)
2022-11-22 11:33:24     at Module.require (node:internal/modules/cjs/loader:1005:19)
2022-11-22 11:33:24     at require (node:internal/modules/cjs/helpers:102:18)
2022-11-22 11:33:24     at Object.<anonymous> (/usr/src/app/dist/src/blocks/blocks.module.js:11:22)
```
- Frontend:
```
2022-11-22 11:34:40 Error: Could not find a production build in the '/usr/src/app/.next' directory. Try building your app with 'next build' before starting the production server. https://nextjs.org/docs/messages/production-start-no-build-id
2022-11-22 11:34:40     at NextNodeServer.getBuildId (/usr/src/app/node_modules/.pnpm/next@12.3.3_m5sxuueb27gk6ddc5gums6vtgq/node_modules/next/dist/server/next-server.js:169:23)
2022-11-22 11:34:40     at new Server (/usr/src/app/node_modules/.pnpm/next@12.3.3_m5sxuueb27gk6ddc5gums6vtgq/node_modules/next/dist/server/base-server.js:58:29)
2022-11-22 11:34:40     at new NextNodeServer (/usr/src/app/node_modules/.pnpm/next@12.3.3_m5sxuueb27gk6ddc5gums6vtgq/node_modules/next/dist/server/next-server.js:70:9)
2022-11-22 11:34:40     at NextServer.createServer (/usr/src/app/node_modules/.pnpm/next@12.3.3_m5sxuueb27gk6ddc5gums6vtgq/node_modules/next/dist/server/next.js:140:16)
2022-11-22 11:34:40     at async /usr/src/app/node_modules/.pnpm/next@12.3.3_m5sxuueb27gk6ddc5gums6vtgq/node_modules/next/dist/server/next.js:149:31
```

Also, if I change `.env` nothing actually changes (even after `source .env`): I tried to change the `WS_PROVIDER` and `FIRST_BLOCK_TO_LOAD` but from the frontend I can't see any changes.

Finally, the blocks stop at ~1417800, is it expected?

**Comment by 0xCaso** [comment]
Oh perfect then, sorry, my fault for the blocks thing. Thank you!

**Comment by 0xslipk** [comment]
@0xCaso we fixed the docker-compose on [main](https://github.com/blockcoders/ink-substrate-explorer-frontend/commit/bd0b2b3153689d3803c2c9c7d7b4024d216117a3) branch, after running `docker-compose up -d` you should be able to access the frontend on `http://localhost:3000/blocks` and the backend on `http://localhost:8080/graphql`

**Comment by 0xCaso** [comment]
Thank you for the changes @0xslipk, now the containers work! By the way, if I change the `WS_PROVIDER` inside the `.env` file, seems like the frontend shows Rococo Contracts parachain data. Could it be because the backend container is still connected to that? The only changing thing seems to be the **Contracts** page (empty). If it's the behavior you wanted it to have, I'll go on with it :)

**Comment by 0xslipk** [comment]
> Thank you for the changes @0xslipk, now the containers work! By the way, if I change the `WS_PROVIDER` inside the `.env` file, seems like the frontend still shows Rococo Contracts parachain data, instead of the local node. Could it be because the backend container is still connected to that? (also, contracts data sometimes show, other times not, do you confirm?)

Awesome. If you change the `WS_PROVIDER` inside the `.env` file you will sync another parachain, so you will need to restart the service. Regarding the contracts data, maybe is because the backend was syncing. We don't have that problem with the [demo](https://ink-explorer.blockcoders.io/contracts) url that is currently synced.

**Comment by 0xCaso** [comment]
Yes, clear, and thank you again. But if I follow these steps:
1. Run my local node with `substrate-contracts-node --dev --ws-external` (or also without `--ws-external`)
2. Edit `WS_PROVIDER` inside `.env` with `ws://127.0.0.1:9944`
3. Run the three containers `backend`, `postgres` and `pgadmin`
4. Start the frontend with `pnpm start:dev`

I still can't see contracts data of the local node, if I upload and interact with contracts both before and after all these 4 steps. I'm using the Substrate Contracts UI, so I'm sure operation with contracts work.

Basically this is the only problem I'm having, could be something on my end but can't understand what.

**Comment by fersirni** [comment]
@0xCaso Hi! I'm reviewing this error you mentioned, I could make it work. I think there might be a miss configuration. 

Can you confirm this?

1. The yaml file has the frontend service commented/deleted
2. You are starting the substrate node locally and NOT adding it to the yaml.
3. In the .env file you are configuring the WS_PROVIDER env var with `ws://<YOUR_IP_ADDRESS>:9944`

This last step won't work if you put localhost, 127.0.0.1 or substrate as this .env is running on a docker container, and the substrate node is running locally on your pc. 
If you start the substrate node using the configuration for the yaml provided on the README then, it will be running inside the docker container hence you can use the alias substrate and should work.

Note: To see contracts on the explorer's 'Contracts' tab the contract has to have events.

If this is not the case and I get the use case wrong please let me know and we'll review it again.
Thanks!

**Comment by 0xCaso** [comment]
Thanks a lot @fersirni, it was because I was using a local node (not with the container) and leaving `127.0.0.1`. I also managed to run it with the node inside the container, as you explained! I'll report these minor things in the evaluations (would be nice to have this explanation in the docs), but they seem to be the only to highlight :)

**Review by 0xCaso — APPROVED** [review]

**Review by Noc2 — APPROVED** [review]

**Comment by 0xslipk** [comment]
> Thanks a lot @fersirni, it was because I was using a local node (not with the container) and leaving `127.0.0.1`. I also managed to run it with the node inside the container, as you explained! I'll report these minor things in the evaluations (it would be nice to have this explanation in the docs), but they seem to be the only to highlight :)

Awesome, Thank you. We are gonna add that in the documentation.

**Comment by 0xCaso** [comment]
@0xslipk @fersirni As you have seen, the delivery has been accepted, thanks again and congrats for the awesome work, keep it up!

I notified the finance team to pay out your invoice. As usual, please allow for up to 14 days for processing.

**Comment by RouvenP** [comment]
hi @0xslipk we transferred the payment today.
